### PR TITLE
Add a Codec for the chunked transfer coding.

### PR DESCRIPTION
**Pull request by nex3**
Related to dart-lang/shelf#64

### CODE REVIEW

File: lib/src/chunked_coding/encoder.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+///
+/// If [isLast] is `true`, this adds the footer that indicates that the chunked
+/// message is complete.
+List<int> _convert(List<int> bytes, int start, int end, {bool isLast: false}) {
+  if (end == start) return isLast ? _doneChunk : const [];
+
+  var size = end - start;
+  var sizeInHex = size.toRadixString(16);
+  var footerSize = isLast ? _doneChunk.length : 0;
+  var list = new Uint8List(sizeInHex.length + 2 + size + footerSize);
```
Comment (efortuna) [inline lib/src/chunked_coding/encoder.dart]: Why +2 here and line 65? 
Consider making a descriptive variable name to clarify

File: lib/src/chunked_coding/decoder.dart [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+          _size = _digitForByte(bytes, start);
+          _state = _State.size;
+          start++;
+          break;
+
+        case _State.size:
+          if (bytes[start] == $cr) {
+            _state = _State.beforeLF;
+          } else {
+            _size = (_size << 4) + _digitForByte(bytes, start);
```
Comment (efortuna) [inline lib/src/chunked_coding/decoder.dart]: consider making 4 a named constant so we don't just have a magic number here.

File: lib/src/chunked_coding.dart [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+export 'chunked_coding/encoder.dart' hide chunkedCodingEncoder;
+export 'chunked_coding/decoder.dart' hide chunkedCodingDecoder;
+
+/// The canonical instance of [ChunkedCodec].
+const chunkedCoding = const ChunkedCodingCodec._();
+
+/// A codec that encodes and decodes the [chunked transfer coding][].
+///
+/// [chunked transfer coding]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1
```
Comment (efortuna) [inline lib/src/chunked_coding.dart]: Why not make this a regular markdown link? 
[chunked transfer coding](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1)

or is this some syntax I'm just behind the times on?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,10 @@
+## 3.1.0
+
+* Add `chunkedCoding`, a `Codec` that supports encoding and decoding the
+  [chunked transfer coding][].
+
+[chunked transfer coding]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1
```
Comment (efortuna) [inline CHANGELOG.md]: Make it a Markdown link? 

[chunked transfer coding](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1)

File: lib/src/chunked_coding/encoder.dart (hunk 1 above)
Comment (nex3) [inline lib/src/chunked_coding/encoder.dart]: Added a comment.

File: CHANGELOG.md (hunk 4 above)
Comment (nex3) [inline CHANGELOG.md]: This is a Markdown link 😊. Search for "implicit link name" in the [syntax documentation](https://daringfireball.net/projects/markdown/syntax#link).

File: lib/src/chunked_coding.dart (hunk 3 above)
Comment (nex3) [inline lib/src/chunked_coding.dart]: See above.

File: lib/src/chunked_coding/decoder.dart (hunk 2 above)
Comment (nex3) [inline lib/src/chunked_coding/decoder.dart]: Added a comment.

File: lib/src/chunked_coding.dart (hunk 3 above)
Comment (efortuna) [inline lib/src/chunked_coding.dart]: doh. :-P

**Pull request merged**